api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

```diff
@@ -535,15 +535,14 @@ public void purgeUser(User user, boolean cascade) throws APIException {
 	 */
 	private void checkPrivileges(Role role) {
 		Collection<Privilege> privileges = role.getPrivileges();
-		
-		if (privileges != null) {
-			for (Privilege p : privileges) {
-				if (!Context.hasPrivilege(p.getPrivilege())) {
-					throw new APIAuthenticationException("Privilege required: " + p);
-				}
-			}
-		}
-	}
+
+		Optional.ofNullable(role.getPrivileges())
+				.map(p -> p.stream().filter(pr -> !Context.hasPrivilege(pr.getPrivilege())).collect(Collectors.toSet()))
+				.ifPresent(missing -> {
+					throw new APIAuthenticationException("Role.you.must.have.privileges: "
```

Comment:
Also the full thing can be re-written as:

```java
Optional.ofNullable(role.getPrivileges())
			.map(p -> p.stream().filter(pr -> !Context.hasPrivilege(pr.getPrivilege())).map(Privilege::getPrivilege)
				.distinct().collect(Collectors.joining(", ")))
				.ifPresent(missing -> {
					throw new APIException("Role.you.must.have.privileges: ", new Object[] { missing });
				});
```

Which is preferable since we only have one call to `Collectors` and one call to `stream()` (in the streaming API, these are the expensive operations which actual result in computation being done).

Reply:
ok thanks let me do that right now